### PR TITLE
[jvm-packages] Fixed test/train persistence

### LINKED ISSUE #2945: [jvm-packages] train and test metrics are exactly the same even when trainTestRatio is set to 0.75 when training with spark
## Environment info
Operating System: Mac

Compiler:
gcc --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 8.1.0 (clang-802.0.42)
Target: x86_64-apple-darwin16.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

Package used (python/R/jvm/C++):JVM

`xgboost` version used: 0.7

## Steps to reproduce

I am call training the data using spark with following code:

    val trainedModel = XGBoost.trainWithDataFrame(
      trainingData,
      trainingParams,
      // num of trees in the model
      round = treeNum,
      // the number of xgboost workers, make it equals to number of trainingData RDD
      // so that when doing the group ranking, the groupData won't be messed up
      nWorkers = partitionNum,
      useExternalMemory = true)
with following default params

  // default training parameters
  val DEFAULT_TRAIN_PARAM: Map[String, Any] = List(
    "eta" -> 0.1f,
    "max_depth" -> 10,
    "silent" -> 1,
    "objective" -> "reg:linear",
    "trainTestRatio" -> "0.75"
  ).toMap

But in the training output, the train_rmse and test_rmse are always the same no matter how many rounds it runs
[INFO] [12/12/2017 17:27:40.819] [RabitTracker-akka.actor.default-dispatcher-9] [akka://RabitTracker/user/Handler] [962]	train-rmse:0.087132	test-rmse:0.087132
[INFO] [12/12/2017 17:27:40.828] [RabitTracker-akka.actor.default-dispatcher-9] [akka://RabitTracker/user/Handler] [963]	train-rmse:0.087132	test-rmse:0.087132
[INFO] [12/12/2017 17:27:40.838] [RabitTracker-akka.actor.default-dispatcher-9] [akka://RabitTracker/user/Handler] [964]	train-rmse:0.087131	test-rmse:0.087131
[INFO] [12/12/2017 17:27:40.847] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [965]	train-rmse:0.087131	test-rmse:0.087131
[INFO] [12/12/2017 17:27:40.859] [RabitTracker-akka.actor.default-dispatcher-11] [akka://RabitTracker/user/Handler] [966]	train-rmse:0.087130	test-rmse:0.087130
[INFO] [12/12/2017 17:27:40.874] [RabitTracker-akka.actor.default-dispatcher-9] [akka://RabitTracker/user/Handler] [967]	train-rmse:0.087130	test-rmse:0.087130
[INFO] [12/12/2017 17:27:40.885] [RabitTracker-akka.actor.default-dispatcher-11] [akka://RabitTracker/user/Handler] [968]	train-rmse:0.087129	test-rmse:0.087129
[INFO] [12/12/2017 17:27:40.894] [RabitTracker-akka.actor.default-dispatcher-9] [akka://RabitTracker/user/Handler] [969]	train-rmse:0.087129	test-rmse:0.087129
[INFO] [12/12/2017 17:27:40.904] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [970]	train-rmse:0.087128	test-rmse:0.087128
[INFO] [12/12/2017 17:27:40.914] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [971]	train-rmse:0.087127	test-rmse:0.087127
[INFO] [12/12/2017 17:27:40.926] [RabitTracker-akka.actor.default-dispatcher-9] [akka://RabitTracker/user/Handler] [972]	train-rmse:0.087126	test-rmse:0.087126
[INFO] [12/12/2017 17:27:40.942] [RabitTracker-akka.actor.default-dispatcher-10] [akka://RabitTracker/user/Handler] [973]	train-rmse:0.087125	test-rmse:0.087125
[INFO] [12/12/2017 17:27:40.952] [RabitTracker-akka.actor.default-dispatcher-10] [akka://RabitTracker/user/Handler] [974]	train-rmse:0.087124	test-rmse:0.087124
[INFO] [12/12/2017 17:27:40.962] [RabitTracker-akka.actor.default-dispatcher-10] [akka://RabitTracker/user/Handler] [975]	train-rmse:0.087123	test-rmse:0.087123
[INFO] [12/12/2017 17:27:40.974] [RabitTracker-akka.actor.default-dispatcher-11] [akka://RabitTracker/user/Handler] [976]	train-rmse:0.087123	test-rmse:0.087123
[INFO] [12/12/2017 17:27:40.983] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [977]	train-rmse:0.087123	test-rmse:0.087123
[INFO] [12/12/2017 17:27:40.998] [RabitTracker-akka.actor.default-dispatcher-10] [akka://RabitTracker/user/Handler] [978]	train-rmse:0.087122	test-rmse:0.087122
[INFO] [12/12/2017 17:27:41.011] [RabitTracker-akka.actor.default-dispatcher-10] [akka://RabitTracker/user/Handler] [979]	train-rmse:0.087121	test-rmse:0.087121
[INFO] [12/12/2017 17:27:41.022] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [980]	train-rmse:0.087120	test-rmse:0.087120
[INFO] [12/12/2017 17:27:41.032] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [981]	train-rmse:0.087118	test-rmse:0.087118
[INFO] [12/12/2017 17:27:41.041] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [982]	train-rmse:0.087117	test-rmse:0.087117
[INFO] [12/12/2017 17:27:41.051] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [983]	train-rmse:0.087117	test-rmse:0.087117
[INFO] [12/12/2017 17:27:41.068] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [984]	train-rmse:0.087116	test-rmse:0.087116
[INFO] [12/12/2017 17:27:41.080] [RabitTracker-akka.actor.default-dispatcher-9] [akka://RabitTracker/user/Handler] [985]	train-rmse:0.087115	test-rmse:0.087115
[INFO] [12/12/2017 17:27:41.089] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [986]	train-rmse:0.087113	test-rmse:0.087113
[INFO] [12/12/2017 17:27:41.100] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [987]	train-rmse:0.087112	test-rmse:0.087112
[INFO] [12/12/2017 17:27:41.115] [RabitTracker-akka.actor.default-dispatcher-10] [akka://RabitTracker/user/Handler] [988]	train-rmse:0.087111	test-rmse:0.087111
[INFO] [12/12/2017 17:27:41.133] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [989]	train-rmse:0.087111	test-rmse:0.087111
[INFO] [12/12/2017 17:27:41.152] [RabitTracker-akka.actor.default-dispatcher-11] [akka://RabitTracker/user/Handler] [990]	train-rmse:0.087110	test-rmse:0.087110
[INFO] [12/12/2017 17:27:41.163] [RabitTracker-akka.actor.default-dispatcher-10] [akka://RabitTracker/user/Handler] [991]	train-rmse:0.087110	test-rmse:0.087110
[INFO] [12/12/2017 17:27:41.174] [RabitTracker-akka.actor.default-dispatcher-11] [akka://RabitTracker/user/Handler] [992]	train-rmse:0.087109	test-rmse:0.087109
[INFO] [12/12/2017 17:27:41.184] [RabitTracker-akka.actor.default-dispatcher-11] [akka://RabitTracker/user/Handler] [993]	train-rmse:0.087108	test-rmse:0.087108
[INFO] [12/12/2017 17:27:41.193] [RabitTracker-akka.actor.default-dispatcher-11] [akka://RabitTracker/user/Handler] [994]	train-rmse:0.087108	test-rmse:0.087108
[INFO] [12/12/2017 17:27:41.211] [RabitTracker-akka.actor.default-dispatcher-9] [akka://RabitTracker/user/Handler] [995]	train-rmse:0.087107	test-rmse:0.087107
[INFO] [12/12/2017 17:27:41.222] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [996]	train-rmse:0.087107	test-rmse:0.087107
[INFO] [12/12/2017 17:27:41.233] [RabitTracker-akka.actor.default-dispatcher-9] [akka://RabitTracker/user/Handler] [997]	train-rmse:0.087106	test-rmse:0.087106
[INFO] [12/12/2017 17:27:41.242] [RabitTracker-akka.actor.default-dispatcher-11] [akka://RabitTracker/user/Handler] [998]	train-rmse:0.087105	test-rmse:0.087105
[INFO] [12/12/2017 17:27:41.252] [RabitTracker-akka.actor.default-dispatcher-2] [akka://RabitTracker/user/Handler] [999]	train-rmse:0.087104	test-rmse:0.087104
[INFO] [12/12/2017 17:27:41.256] [RabitTracker-akka.actor.default-dispatcher-11] [akka://RabitTracker/user/Handler] Received shutdown signal from 0
17/12/12 17:28:16 INFO MemoryStore: Block rdd_9_0 stored as values in memory (estimated size 8.1 KB, free 908.2 MB)

## What have you tried?

1. I debug the code step by step and I can see in following code, testPoints and trainPoints are correctly generated.

private object Watches {

  def apply(
      params: Map[String, Any],
      labeledPoints: Iterator[XGBLabeledPoint],
      baseMarginsOpt: Option[Array[Float]],
      cacheFileName: String): Watches = {
    val trainTestRatio = params.get("trainTestRatio").map(_.toString.toDouble).getOrElse(1.0)
    val seed = params.get("seed").map(_.toString.toLong).getOrElse(System.nanoTime())
    val r = new Random(seed)
    val testPoints = mutable.ArrayBuffer.empty[XGBLabeledPoint]
    val trainPoints = labeledPoints.filter { labeledPoint =>
      val accepted = r.nextDouble() <= trainTestRatio
      if (!accepted) {
        testPoints += labeledPoint
      }

      accepted
    }
    val trainMatrix = new DMatrix(trainPoints, cacheFileName)
    val testMatrix = new DMatrix(testPoints.iterator, cacheFileName)
    r.setSeed(seed)
    for (baseMargins <- baseMarginsOpt) {
      val (trainMargin, testMargin) = baseMargins.partition(_ => r.nextDouble() <= trainTestRatio)
      trainMatrix.setBaseMargin(trainMargin)
      testMatrix.setBaseMargin(testMargin)
    }

    // TODO: use group attribute from the points.
    if (params.contains("groupData") && params("groupData") != null) {
      trainMatrix.setGroup(params("groupData").asInstanceOf[Seq[Seq[Int]]](
        TaskContext.getPartitionId()).toArray)
    }
    new Watches(train = trainMatrix, test = testMatrix)
  }
}

### PR DESCRIPTION
**Pull request by superbobry**
Prior to this PR both data sets were persisted in the same directory, i.e. the test data replaced the training one which led to

* training on less data (since usually test < train) and
* test loss being exactly equal to the training loss.

Closes #2945.

### CODE REVIEW

**Comment by CodingCat** [comment]
I will review it tmr but the test is failed because cmake uses too much memory?

**Comment by CodingCat** [comment]
I just merged a change which disabled their annoying image change....you can try to rebase with master

**Comment by superbobry** [comment]
I wonder if the test failures are due to dmlc/dmlc-core#323...

**Comment by CodingCat** [comment]
let me know when it is ready for review

**Comment by superbobry** [comment]
@CodingCat should be ready.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 1]
```diff
@@ -122,17 +124,21 @@ object XGBoost extends Serializable {
           s"detected an empty partition in the training data, partition ID:" +
             s" ${TaskContext.getPartitionId()}")
       }
-      val cacheFileName = if (useExternalMemory) {
-        s"$appName-${TaskContext.get().stageId()}-" +
-          s"dtrain_cache-${TaskContext.getPartitionId()}"
+      val taskId = TaskContext.getPartitionId().toString
+      val cacheDirName = if (useExternalMemory) {
+        val dir = new File(s"$appName-${TaskContext.get().stageId()}-dtrain_cache-$taskId")
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: replace dtrain_cache with cache

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 2]
```diff
@@ -145,6 +151,12 @@ object XGBoost extends Serializable {
       } finally {
         Rabit.shutdown()
         watches.delete()
+        cacheDirName.foreach { name =>
+          val cacheDir = new File(name)
+          if (!cacheDir.listFiles().forall(_.delete()) || !cacheDir.delete()) {
+            throw new IllegalStateException(s"failed to delete $cacheDir")
+          }
+        }
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: shouldn't the cache files be created under executor's working directory which would be cleaned by Spark automatically?

Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: I think `cacheDirName` is inside the working directory since it's a relative path. I would rather do the cleanup manually because otherwise, we could fail the container due to consuming all available disk space (for no good reason).

Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: We could reduce the length of the filename though, taskId should be unique enough.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 1 above)
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Will do.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 2 above)
Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: it's possible that the user will create multiple models in the same spark application, so I would like to keep stageId

Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: I think for cleaning, at least we should do in watches.delete() instead of leaking to main training logic

Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: The reason I did not include it in `watches.delete` is that `Watches` does not create the directory, so imo it shouldn't remove it. However, it can remove the individual files for the matrices. Sounds good?

Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: it's fine, watches didn't create DMatrix either but still cleaned it....

**Comment by CodingCat** [comment]
LGTM
